---
title: "tl21-01_pirsInTheDeNovoAssemblies"
output: 
  html_document:
    theme: cosmo
    code_download: true
    toc: true
    number_sections: true
    toc_float:
      collapsed: false
      smooth_scroll: true
params:
  blast_output_file: '/camp/lab/langhornej/working/HPC/littlet/tl21-01_testingTrinityChab24h/blast_pirDatabase/scspades/scspades_SBPrMT-mR-rRNAdep/scspades_SBPrMT-mR-rRNAdep_blastn.outfmt6'
---

# Intro

This script is looks at the BLAST results of a Trinity _de novo_ assembled transcriptome. 

# What this report shows:

*The level of counts / TPM of pirs that are detected by the Trinity de novo assembled transcriptome.
**e.g. What is the minimum expressed gene detected / What is the maximally expressed gene not detected.

# To-do:
*Check how they BLAST to the rest of the transcriptome, do pir de novo transcripts also BLAST to non-pir transcripts?
*Select the most informative data and poo-poo the rest.
*Save useful files made by this script as csv.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, 
                      results = TRUE, 
                      message = FALSE, 
                      warning = FALSE)
options(scipen = 999) # Means that numbers are displayed normally not as 1e6 (1000000)

#Possible files for blast_output_file:
# '/camp/lab/langhornej/working/HPC/littlet/tl21-01_testingTrinityChab24h/blast_pirDatabase/Trinity/trinityAll_blast_211129/trinityDeNovo_PirTranscriptsPccAS_blastn.outfmt6'
# '/camp/lab/langhornej/working/HPC/littlet/tl21-01_testingTrinityChab24h/blast_pirDatabase/Trinity/trinity_blast_211124/trinityDeNovo_PirTranscriptsPccAS_blastn.outfmt6'
```

```{r loading_packages, include = FALSE, message = FALSE}
# Load the required packages
library('readr')
library('dplyr', 
        lib.loc = "/camp/home/littlet/R/x86_64-pc-linux-gnu-library/4.0")
library('stringr')
library('purrr')
library('IRanges')
library('kableExtra')
library('plotly')
```

```{r function_ordering}
ordering <- function(to_order, order_vector, fromLast = TRUE){
  order = unlist(sapply(order_vector, 
                        function(x){unique(str_subset(to_order, 
                                                      pattern = paste(x)))}))
  order_missing = unique(to_order[!to_order %in% order])
  unique(c(order, order_missing), fromLast = fromLast)
}
```

```{r function_plotRanges}
#' Plot genomic ranges from the IRanges object or GRanges object.
# from https://rdrr.io/github/JokingHero/rxtras/src/R/plot_wrappers.R

plotRanges <- function(x, xlim = x, main = deparse(substitute(x)),  col = "black", sep = 0.5, ...){
  if (!requireNamespace("IRanges", quietly = TRUE)) {
    stop("IRanges needed for this function to work. Please install it.",
         call. = FALSE)
  }
  height <- 1
  if (is(xlim, "Ranges"))
      xlim <- c(min(start(xlim)), max(end(xlim)))
  bins <- disjointBins(IRanges(start(x), end(x) + 1))
  plot.new()
  plot.window(xlim, c(0, max(bins) * (height + sep)))
  ybottom <- bins * (sep + height) - height
  rect(start(x) - 0.5, ybottom, end(x) + 0.5, ybottom + height, col = col, ...)
  title(main)
  axis(1)
}
```

```{r load_data}
pir.info <- read.delim('Pchabaudi_pir_info.txt', sep = '\t')

counts <- read.csv('tl19-07_twentyfour_counts_samples.csv',
                   header = TRUE,
                   stringsAsFactors = FALSE) %>% 
  filter(Geneid %in% pir.info$Gene.ID) %>% 
  select(colnames(.)[order(colnames(.))])

tpm_data <- read.csv('tl19-07_twentyfour_tpm_samples.csv',
                   header = TRUE,
                   stringsAsFactors = FALSE) %>% 
  filter(Geneid %in% pir.info$Gene.ID) %>% 
  select(colnames(counts)[order(colnames(counts))])
```

```{r read_blast_results}
# list.files('../blast', pattern = 'outfmt')

#Set a filter for bp length to assume is a valid match
filter_nident <- 100

outfmt6_headers <- 
  'qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp slen nident' %>% 
  strsplit(split = ' ') %>% 
  unlist

blast <- read.table(params$blast_output_file,
                    col.names = outfmt6_headers) %>% 
  mutate(geneid = str_extract(sseqid, 
                              pattern = '(?<=gene=).+(?=___or)'),
         sseqid = unlist(
           lapply(
             str_extract_all(sseqid, 
                             pattern = '(?<=gene=).+(?=___or)|(?<=gene_product=).+(?=___trans)'), 
             paste, 
             collapse = '_')
         ),
         scovhsp = round( (nident/slen * 100), digits = 2),
         scovs = round(((abs(send - sstart) + 1)/slen *100), digits = 2),
         evalue = format(evalue, scientific = TRUE)
  ) %>% 
  filter(nident > filter_nident) %>%
  group_by(qseqid) %>% 
  mutate(num_genes_match = n_distinct(geneid))

# Filter for the top hits for each de novo transcript by e-value.

blast_top <- group_by(blast, qseqid) %>% 
  slice_min(evalue, n = 1)

#Filter for the de novo transcripts which only match a single known pir.
# Also removing idents below 100 because these are probably poor matches.

blast_uniq <- filter(blast, num_genes_match == 1)

#sanity test
# all(blast$scovs >= blast$scovhsp)
```

Filtering matches by a minimum of `r filter_nident`bp to be a valid match.

# Number of hits for each de novo transcript with a BLAST hit.

Are there many de novo transcripts which match multiple different known transcripts? Are most of them unique?

```{r number_of_hits_per_deNovoTranscript}
blast %>% 
  group_by(qseqid) %>% 
  summarise(num = n_distinct(geneid)) %>% 
  .$num %>% 
  hist(main = 'Histogram of BLAST hits per de novo transcript (50 max)',
       breaks = 50)
```

#Genes with the most number of uniquely mapping de novo transcripts.

```{r number_of_uniq_transcripts_per_gene}
blast %>% 
  filter(num_genes_match== 1) %>% 
  group_by(geneid) %>% 
  summarise(num_of_uniq_deNovoTranscripts = n_distinct(qseqid),
            max_id = max(pident),
            min_id = min(pident),
            max_nident = max(nident),
            min_nident = min(nident),
            max_gap = max(gapopen),
            min_gap = min(gapopen),
            max_qcovh = max(qcovhsp),
            min_qcovh = min(qcovhsp),
            max_scovh = max(scovhsp),
            min_scovh = min(scovhsp)) %>% 
  arrange(desc(num_of_uniq_deNovoTranscripts)) %>% 
  kbl() %>% kable_styling() %>% scroll_box(width = "100%", height = "400px")
```

# Coverage of the subject (known transcripts) and query (assembled de novo transcripts) sequences. 

Using only the de novo transcripts that uniquely match one known pir transcript.

For our purposes the subject coverage is probably most important.

```{r scov_v_qcov}
fig <- plot_ly(
  blast_uniq,
  x = ~qcovhsp,
  y = ~scovhsp,
  text = ~paste(sseqid)
) %>% 
  layout(xaxis = list(title = "De novo transcript coverage"),
         yaxis = list(title = "Known transcript coverage"))
fig
```

Number of unique genes found by BLAST: `r unique(blast$sseqid) %>% length`.

# Known transcripts with >95% coverage from one de novo transcript.

How many of the known transcripts are covered >=95% by de novo transcripts?

num_match_all and num_match_95 show how many other known transcripts these de novo transcripts match in total or >=95% respectively.

```{r transcripts_covering_most_of_known_transcripts}
transcripts_covered_by_denovo <- blast %>% 
  group_by(qseqid) %>%
  mutate(num_match_all = n_distinct(sseqid)) %>%
  group_by(sseqid) %>% 
  filter(scovhsp >= 95) %>% 
  group_by(qseqid) %>%
  mutate(num_match_95 = n_distinct(sseqid)) %>% 
  arrange(geneid)
transcripts_covered_by_denovo %>% 
  select(-qstart,-qend,-sstart,-send,-evalue,-bitscore,-qcovs,-slen,-geneid,-nident,-scovs) %>% 
  kbl() %>% kable_styling() %>% scroll_box(width = "100%", height = "300px")
```

Showing the BLAST alignments of these >=95% coverage de novo transcripts in a graph, and all the matches they have. 
The "n =" number after the transcript name shows how many other BLAST hits the de novo transcript had in total.
Size of each point shows the length of the alignment.

```{r transcripts_covering_most_of_known_transcripts_graph}
blast %>% 
  filter(qseqid %in% unique(transcripts_covered_by_denovo$qseqid)) %>%
  group_by(qseqid) %>%
  mutate(num_match_all = n_distinct(sseqid)) %>% 
  mutate(qseqid = paste0(qseqid, '; n = ', num_match_all),
         over95 = ifelse(scovhsp >= 95, 
                         '>=95% coverage', 
                         '<95% coverage')) %>% 
  plot_ly(x = ~qcovhsp,
          y = ~scovhsp,
          text = ~paste(sseqid, '\n', 
                        length, 'bp', '\n',
                        scovhsp, '% coverage of known'),
          color = ~qseqid,
          size = ~length,
          symbol = ~over95,
          symbols = c('triangle-up', 'circle'),
          sizes = c(2,100),
          legendgroup = ~qseqid) %>% 
  layout(xaxis = list(title = "De novo transcript coverage"),
         yaxis = list(title = "Known transcript coverage"))
```

Alignments with >=95% known coverage but <50% de novo transcript coverage. 

```{r low_qcovhsp}
transcripts_covered_by_denovo %>% 
  filter(qcovhsp < 50) %>% 
  kbl() %>% kable_styling() %>% scroll_box(width = "100%", height = "300px")
```

# Transcripts matched by percentage coverage.

```{r transcript_coverage_by_percentage}

transcripts_covered_by_denovo_diffcov <- lapply(seq(0,100,by = 5),
                                                function(covrg) {
                                                  blast %>% 
                                                    group_by(sseqid) %>% 
                                                    filter(scovhsp >= covrg) %>%
                                                    .$geneid %>% unique %>% length
                                                })
names(transcripts_covered_by_denovo_diffcov) <- seq(0,100,by = 5)
plot(y = unlist(transcripts_covered_by_denovo_diffcov), 
     x = seq(0,100,by = 5),
     ylim = c(0,225),
     main = 'For all de novo transcripts',
     ylab = 'Number of known transcripts covered',
     xlab = 'Percentage of coverage by individual transcripts',
     yaxt = 'none')
axis(2, at = seq(0,250,by = 25))
abline(a = 25, b = 0)
                                                    
```

```{r uniq_transcript_coverage_by_percentage}
transcripts_covered_by_denovo_uniq_diffcov <- lapply(seq(0,100,by = 5),
                                                function(covrg) {
                                                  blast_uniq %>% 
                                                    group_by(sseqid) %>% 
                                                    filter(scovhsp >= covrg) %>%
                                                    .$geneid %>% unique %>% length
                                                })
plot(y = unlist(transcripts_covered_by_denovo_uniq_diffcov), 
     x = seq(0,100,by = 5),
     ylim = c(0,225),
     main = 'For uniquely mapping de novo transcipts',
     ylab = 'Number of known transcripts covered',
     xlab = 'Percentage of coverage by individual transcripts',
     yaxt = 'none')
axis(2, at = seq(0,250,by = 25))
abline(a = 25, b = 0)
                   
```

# Highest coverage by an individual transcript v counts/TPM

```{r high_cov_v_counts}
blast_high <- blast %>% 
  group_by(sseqid, geneid) %>% 
  slice_min(order_by = evalue, n = 1) %>% 
  slice_max(order_by = scovhsp, n = 1)

length(unique(blast_high$qseqid)) # Some may be shared between genes.
```


# Number of top hits for each de novo transcript v counts/TPM

```{r uniq_hits_v_counts}
num_uniq_hits <- blast_top %>% 
  group_by(geneid) %>% 
  summarise(num = n()) %>% 
  arrange(geneid) %>% 
  filter(geneid %in% counts$Geneid)

total_counts_ofBlast <- filter(counts, 
                               Geneid %in% pir.info$Gene.ID) %>% 
  rowwise() %>% 
  mutate(geneid = Geneid,
  total_counts = sum(c_across(matches('SBP|RMT')))) %>% 
  select(c('geneid','total_counts'))

#Missing pir genes set to zero 
missing <- tibble(geneid = pir.info$Gene.ID[!pir.info$Gene.ID %in% num_uniq_hits$geneid],
                  num = 0)

num_uniq_hits <- rbind(num_uniq_hits,
                       missing) %>% 
  arrange(geneid)

total_tpm_ofBlast <- filter(tpm_data, 
       Geneid %in% pir.info$Gene.ID) %>% 
  arrange(Geneid) %>% 
  rowwise() %>% 
  mutate(geneid = Geneid, 
         total_tpm = sum(c_across(matches('SBP|rMT')))) %>% 
  select(c('geneid','total_tpm'))

# Combine the dfs

gene_transcription_df <- merge(merge(num_uniq_hits, 
                                     total_tpm_ofBlast),
                               total_counts_ofBlast)

#all(num_uniq_hits$geneid %in% total_counts_ofBlast$Geneid)

plot(log1p(gene_transcription_df$num), 
     log1p(gene_transcription_df$total_counts),
     xlab = 'Number of de novo transcripts which BLAST to the known transcript best',
     ylab = 'log Total gene counts + 1',
     xaxt = "n")
axis(1, at = log1p(c(0, 1, 2, 3, 5, 20, 50, 100)), labels = c(0, 1, 2, 3, 5, 20, 50, 100))

plot(log1p(gene_transcription_df$num), 
     log1p(gene_transcription_df$total_tpm),
     xlab = 'Number of de novo transcripts which BLAST to the known transcript best',
     ylab = 'log Total gene tpm + 1',
     xaxt = "n")
axis(1, at = log1p(c(0, 1, 2, 3, 5, 20, 50, 100)), labels = c(0, 1, 2, 3, 5, 20, 50, 100))

```
```{r cov_by_all_hits}
# Percentage of the subject covered by all hits.
#  Need to make sure the order is always the same.
#  I think this coverage could still include gaps.

blast_dir <- blast %>% 
  mutate(direction = ifelse(sstart > send, '-', '+'),
         sstart_dir = ifelse(sstart > send, send, sstart),
         send_dir = ifelse(sstart > send, sstart, send))

gene_cov_all_hits <- IRanges(blast_dir$sstart_dir, blast_dir$send_dir) %>% 
  split(., blast_dir$geneid) %>% 
  reduce() %>% 
  width %>% 
  sum %>% 
  data.frame

gene_cov_all_hits$length <- blast$slen[match(rownames(gene_cov_all_hits), 
                                  blast$geneid)]
gene_cov_all_hits$qcovall <- 100*gene_cov_all_hits$./gene_cov_all_hits$length
gene_cov_all_hits$geneid <- rownames(gene_cov_all_hits)

gene_cov_all_hits <- merge(merge(gene_cov_all_hits, 
                                 total_tpm_ofBlast),
                           total_counts_ofBlast)

plot(gene_cov_all_hits$qcovall, 
     log1p(gene_cov_all_hits$total_counts),
     xlab = 'Total coverage of known transcript by all de novo transcripts',
     ylab = 'log total counts + 1')

```

# Are the genes not found by Trinity detected with lower count values than those in the assembly?

```{r counts_of_hits_v_nohits}
num_uniq_hits <- blast_top %>% 
  group_by(sseqid, geneid) %>% 
  summarise(num = n()) %>% 
  mutate(top_hits = ifelse(num >= 1, 'Yes', 'No')) %>% 
  arrange(geneid) %>% 
  filter(geneid %in% counts$Geneid)

total_counts <- counts %>%  
  rowwise() %>% 
  mutate(total_counts = sum(c_across(matches('SBP|rMT')))) %>% 
  ungroup %>% 
  select(c('Geneid','total_counts'))

total_counts_BlastHits <- mutate(total_counts,
       top_hits = ifelse(Geneid %in% num_uniq_hits$geneid, 'Yes', 'No'))

total_tpm <- tpm_data %>%  
  rowwise() %>% 
  mutate(total_tpm = sum(c_across(matches('SBP|rMT')))) %>% 
  ungroup %>% 
  select(c('Geneid','total_tpm'))

total_tpm_BlastHits <- mutate(total_tpm,
       top_hits = ifelse(Geneid %in% num_uniq_hits$geneid, 'Yes', 'No'))

boxplot(log1p(total_counts) ~ top_hits , data = total_counts_BlastHits)
# boxplot(log1p(total_tpm) ~ top_hits , data = total_tpm_BlastHits)
```

#What's the lowest count/tpm assembled gene and the highest count/tpm undetected gene?

```{r high_count_unassembled}
total_counts_BlastHits %>% 
  group_by(top_hits) %>% 
  filter(top_hits == 'No') %>% 
  dplyr::slice_max(total_counts, n = 5) %>%  
  kbl() %>% kable_styling() %>% scroll_box(width = "100%", height = "200px")

total_tpm_BlastHits %>%
  group_by(top_hits) %>%
  filter(top_hits == 'No') %>% 
  dplyr::slice_max(total_tpm, n = 5) %>%  
  kbl() %>% kable_styling() %>% scroll_box(width = "100%", height = "200px")
```

```{r low_count_assembled}
low_count_hit <- total_counts_BlastHits %>% 
  group_by(top_hits) %>% 
  dplyr::slice_min(total_counts, n = 5) %>% 
  filter(top_hits == 'Yes')

low_count_hit %>%  
  kbl() %>% kable_styling() %>% scroll_box(width = "100%", height = "200px")

IRanges(blast_dir$sstart_dir, blast_dir$send_dir) %>% 
  split(., blast_dir$geneid) %>% 
  `[[`(., paste(low_count_hit[1,1])) %>% 
  plotRanges(main = paste(low_count_hit[1,1]))

low_tpm_hit <- total_tpm_BlastHits %>% 
  group_by(top_hits) %>% 
  dplyr::slice_min(total_tpm, n = 5) %>% 
  filter(top_hits == 'Yes')

low_tpm_hit %>%  
  kbl() %>% kable_styling() %>% scroll_box(width = "100%", height = "200px")

IRanges(blast_dir$sstart_dir, blast_dir$send_dir) %>% 
  split(., blast_dir$geneid) %>% 
  `[[`(., paste(low_tpm_hit[1,1])) %>% 
  plotRanges(main = paste(low_tpm_hit[1,1]))
```

```{r assemblies_of_low_counts}
genes_low_count_assembled <- total_counts_BlastHits %>% 
  group_by(top_hits) %>% 
  dplyr::slice_min(total_counts, n = 5) %>% 
  filter(top_hits == 'Yes') %>% 
  .$Geneid

blast %>% 
  filter(str_detect(sseqid, 
                    pattern = paste(genes_low_count_assembled, 
                                    collapse = '|'))) %>% 
  arrange(sseqid, evalue) %>%  
  kbl() %>% kable_styling() %>% scroll_box(width = "100%", height = "200px")
```

# Known transcripts with unique and multi-BLASTing de novo transcripts

```{r pirs_with_uniq_transcript_matched}
multi_uniq_transcript <- blast %>% 
  group_by(qseqid) %>% 
  mutate(num_pirs = ifelse(length(unique(sseqid)) == 1, 
                           'Uniq', 
                           'Multimatch')) %>% 
  ungroup() %>% 
  group_by(sseqid, geneid) %>% 
  summarise(uniq_trans = length(unique(qseqid[num_pirs == 'Uniq'])),
            multi_trans = length(unique(qseqid[num_pirs == 'Multimatch'])),
            num_uniq = length(unique(qseqid)),
            num = length(qseqid),
            lens = paste0(max(length), ', ', min(length)),
            pident = paste0(max(pident),', ',min(pident)),
            # transcripts = paste0(unique(qseqid), 
            #                      collapse = ', ')
            ) 

multi_uniq_transcript %>% 
  select(-geneid) %>% 
  arrange(desc(uniq_trans), multi_trans )%>%  
  kbl() %>% kable_styling() %>% scroll_box(width = "100%", height = "400px")
```

## Known transcripts with only multimapping hits.

```{r pirs_with_only_multi_transcript_matched}
multi_uniq_transcript %>% 
  select(-geneid) %>% 
  filter(uniq_trans == 0) %>% 
  arrange(desc(multi_trans) )%>%  
  kbl() %>% kable_styling() %>% scroll_box(width = "100%", height = "300px")
```

## Multimapping transcripts by expression

```{r multimap_expression}
#Note that apparent discrepencies may arise when comparing between these data frames.
# The number of unique or total transcripts will depend on whether
# one is only using the best BLAST hits for each query 
# (e.g. gene_transcription_df), and/or whether the same query with multiple
# alignments to the same subject is included or not.

multimap_exprssn_df <- merge(gene_transcription_df, 
                             multi_uniq_transcript, 
                             by = 'geneid')
plot(multimap_exprssn_df$multi_trans,
     log1p(multimap_exprssn_df$total_counts),
     xlab = 'Number of de novo transcript hits that are \'shared\' with other known transcripts',
     ylab = 'log Total counts + 1')

```


```{r possible_chimeric_recomb_sequences, eval = FALSE}

length_filter <- 200

genes_sharingfewertranscripts <- blast %>% 
  filter(length >= length_filter) %>% 
  group_by(qseqid) %>% 
  mutate(num_genes_hit = n_distinct(sseqid)) %>% 
  group_by(geneid) %>% 
  mutate(num_withotherpirs = ifelse(num_genes_hit >= 2,
                                     1,
                                     0)) %>% 
  distinct(qseqid, sseqid, .keep_all = TRUE) %>% 
  summarise(sum_num_withotherpirs = sum(num_withotherpirs)) %>% 
  arrange(desc(sum_num_withotherpirs)) %>% 
  filter(sum_num_withotherpirs < 10) %>% 
  .$geneid %>% 
  sort

#How many known transcripts hit a de novo transcript which also hits
# other known transcripts (possible chimeras/recombinants, or a result of 
# sequence similarity)? 

# Note that the distinct() function, to remove duplicates of known transcripts which BLAST
#   to the same gene multiple times, makes a big difference.

blast %>% 
  filter(length >= length_filter) %>%
  group_by(qseqid) %>% 
  mutate(num_genes_hit = n_distinct(sseqid)) %>% 
  group_by(geneid) %>% 
  mutate(num_withotherpirs = ifelse(num_genes_hit >= 2,
                                     1,
                                     0)) %>% 
  distinct(qseqid, sseqid, .keep_all = TRUE) %>% 
  summarise(sum_num_withotherpirs = sum(num_withotherpirs)) %>% 
  .$sum_num_withotherpirs %>% 
  hist(breaks = 20, xlab = 'Number of shared de novo transcripts')

blast %>% 
  filter(length >= length_filter,
         geneid %in% genes_sharingfewertranscripts) %>% 
  group_by(qseqid) %>% 
  mutate(num_genes_hit = n_distinct(sseqid)) %>% 
  arrange(desc(num_genes_hit), qseqid)
transcripts_covered_by_denovo_diffcov$`75`
```

# Important instant details:

Filtered matches by a minimum of `r filter_nident`bp to be a valid match.

There are `r length(unique(transcripts_covered_by_denovo$geneid))` known transcripts covered by >=95% of at least one individual _de novo_ transcript. There is a mean of `r mean(transcripts_covered_by_denovo$mismatch)` mismatches and `r mean(transcripts_covered_by_denovo$gapopen)` gaps and a median of `r median(transcripts_covered_by_denovo$mismatch)` mismatches and `r median(transcripts_covered_by_denovo$gapopen)` gaps in these alignments (Note that there are probably some de novo transcripts that align >=95% to the same known transcript). Of these `r length(unique(transcripts_covered_by_denovo$geneid[transcripts_covered_by_denovo$num_match_all == 1]))`
 match only that known transcript and no others by the BLAST search. `r length(unique(filter(ungroup(transcripts_covered_by_denovo), qcovhsp < 50)$geneid))` of the genes with >=95% known transcript alignments cover less than 50% of the de novo transcript.
 
See section [Known transcripts with >95% coverage from one de novo transcript.].
 
 With a threshold of 75% coverage `r transcripts_covered_by_denovo_diffcov['75']` known transcripts are represented by the de novo transcripts.
 
See section [Transcripts matched by percentage coverage.].
