---
title: "tl21-01_testingTrinityChab24h_DetectedTranscriptsByExpression"
output: 
  html_document:
    theme: cosmo
    code_download: true
    toc: true
    number_sections: true
    toc_float:
      collapsed: false
      smooth_scroll: true
---

# Intro

This script is looking at the BLAST results (only reporting the top unique hit) of the first Trinity de novo assembled transcriptome I did, against the transcriptome of P chabaudi chabaudi AS from PlasmoDB v53. 

The de novo assembly used the reads from the 24h SBP asexual blood cycle experiment, which took samples every 3h over an entire blood cycle.

# Aims

* Ascertain the level of counts / TPM that are detected by the Trinity de novo assembled transcriptome.

** e.g. What is the minimum expressed gene detected / What is the maximally expressed gene not detected.

*Same as above but for the _pir_ genes.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, 
                      results = TRUE, 
                      message = FALSE, 
                      warning = FALSE)
options(scipen = 999) # Means that numbers are displayed normally not as 1e6 (1000000)
```

```{r loading_packages, include = FALSE, message = FALSE}
# Load the required packages
library('readr')
library('dplyr', 
        lib.loc = "/camp/home/littlet/R/x86_64-pc-linux-gnu-library/4.0")
library('stringr')
library('purrr')
```

```{r function_ordering}
ordering <- function(to_order, order_vector, fromLast = TRUE){
  order = unlist(sapply(order_vector, 
                        function(x){unique(str_subset(to_order, 
                                                      pattern = paste(x)))}))
  order_missing = unique(to_order[!to_order %in% order])
  unique(c(order, order_missing), fromLast = fromLast)
}
```

```{r load_data}
counts <- read.csv('tl19-07_twentyfour_counts_samples.csv',
                   header = TRUE,
                   stringsAsFactors = FALSE) %>% 
  select(matches('Geneid|SBP')) %>% 
  select(colnames(.)[order(colnames(.))])

sbp <- read.csv('tl19-07_twentyfour_tpm_samples.csv',
                   header = TRUE,
                   stringsAsFactors = FALSE) %>% 
  select(matches('Geneid|SBP')) %>% 
  select(colnames(counts)[order(colnames(counts))])
```

```{r check_tpm}

#Check that TPM all add to a million

# all(
#   sapply(apply(sbp[,-1], 2, sum), 
#          all.equal, 
#          1e6, 
#          tolerance = 0.001)
# )
  
```

```{r read_blast_results}
# list.files('../blast', pattern = 'outfmt')

outfmt6_headers <- 
  'qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore' %>% 
  strsplit(split = ' ') %>% 
  unlist

all_blast <- read.table('../blast/trinityDeNovo_TranscriptsPccAS_blastn.outfmt6',
                        col.names = outfmt6_headers) %>% 
  mutate(geneid = str_extract(sseqid, 
                              pattern = '(?<=gene=).+(?=___or)'),
         sseqid = unlist(
           lapply(
             str_extract_all(sseqid, 
                             pattern = '(?<=gene=).+(?=___or)|(?<=gene_product=).+(?=___trans)'), 
             paste, 
             collapse = '_')
         )
  )

pir_blast <- filter(all_blast,
                    str_detect(sseqid, 
                               pattern = 'PIR_protein(?!.+pseudo)'))

# unique(pir_blast$sseqid) %>% length
```

# Top unique hits across the transcriptome.

```{r top_uniq_hits}
all_blast %>% 
  group_by(sseqid) %>% 
  summarise(num = n()) %>% 
  arrange(desc(num))
```

Interesting that the seventh and eighth of the top unique hits are pirs:
* PCHAS_0301700.
* PCHAS_0701400.

# Top unique 100% identity hits.

```{r top_100identity}
all_blast %>% 
  group_by(sseqid) %>% 
  filter(pident == 100) %>% 
  summarise(num = n()) %>% 
  arrange(desc(num))
```

These are all different from the general top unique hits, that's quite interesting.

# Number unique hits v counts

```{r uniq_hits_v_counts}
num_uniq_hits <- all_blast %>% 
  group_by(sseqid, geneid) %>% 
  summarise(num = n()) %>% 
  arrange(geneid) %>% 
  filter(geneid %in% counts$Geneid)

total_counts_ofBlast <- filter(counts, 
       Geneid %in% num_uniq_hits$geneid) %>% 
  rowwise() %>% 
  mutate(total_counts = sum(c_across(matches('SBP')))) %>% 
  select(c('Geneid','total_counts'))

#all(num_uniq_hits$geneid %in% total_counts_ofBlast$Geneid)

plot(num_uniq_hits$num, total_counts_ofBlast$total_counts)

```

# Are the genes not found by Trinity detected with lower count values than those in the assembly?

```{r}
num_uniq_hits <- all_blast %>% 
  group_by(sseqid, geneid) %>% 
  summarise(num = n()) %>% 
  mutate(any_hits = ifelse(num >= 1, 'Yes', 'No')) %>% 
  arrange(geneid) %>% 
  filter(geneid %in% counts$Geneid)

total_counts <- counts %>%  
  rowwise() %>% 
  mutate(total_counts = sum(c_across(matches('SBP')))) %>% 
  ungroup %>% 
  select(c('Geneid','total_counts'))

total_counts_BlastHits <- mutate(total_counts,
       any_hits = ifelse(Geneid %in% num_uniq_hits$geneid, 'Yes', 'No'))

boxplot(log1p(total_counts) ~ any_hits , data = total_counts_BlastHits)
```

Overall yes but not by much.

```{r wilcox_test}
wilcox.test(total_counts$total_counts[total_counts_BlastHits$any_hits == 'No'], 
            total_counts$total_counts[total_counts_BlastHits$any_hits == 'Yes'])
```

#What's the lowest count assembled gene and the highest count undetected gene?

```{r high_count_unassembled}
total_counts_BlastHits %>% 
  group_by(any_hits) %>% 
  dplyr::slice_max(total_counts, n = 5)
```

```{r low_count_assembled}
total_counts_BlastHits %>% 
  group_by(any_hits) %>% 
  dplyr::slice_min(total_counts, n = 5) %>% 
  map_df(rev) #Turn df upside-down
```

PCHAS_1109820 is the lowest gene count to appear in the Trinity assembly BLAST.
Rather sanely no zero count genes are found by Trinity.

```{r}
genes_low_count_assembled <- total_counts_BlastHits %>% 
  group_by(any_hits) %>% 
  dplyr::slice_min(total_counts, n = 5) %>% 
  filter(any_hits == 'Yes') %>% 
  .$Geneid

all_blast %>% 
  filter(str_detect(sseqid, 
                    pattern = paste(genes_low_count_assembled, 
                                    collapse = '|')))
```

And they're good assemblies!

# _pir_ gene specific

```{r pir_uniq_hits_v_counts}
pir_num_uniq_hits <- pir_blast %>% 
  group_by(sseqid, geneid) %>% 
  summarise(num = n()) %>% 
  arrange(geneid) %>% 
  filter(geneid %in% counts$Geneid)

pir_total_counts_ofBlast <- filter(counts, 
       Geneid %in% pir_num_uniq_hits$geneid) %>% 
  rowwise() %>% 
  mutate(total_counts = sum(c_across(matches('SBP')))) %>% 
  select(c('Geneid','total_counts'))

#all(num_uniq_hits$geneid %in% total_counts_ofBlast$Geneid)

plot(pir_num_uniq_hits$num, pir_total_counts_ofBlast$total_counts)

```

Are two pirs with lots of unique hits but low counts, and two pirs with high counts but low number of unique hits.

# Are the genes not found by Trinity detected with lower count values than those in the assembly?

```{r}
list.pirs <- read_delim('Pchabaudi_pir_info.txt', 
                        delim = '\t')

pir_total_counts <- filter(total_counts,
                           Geneid %in% list.pirs$Gene.ID)

pir_total_counts_BlastHits <- mutate(pir_total_counts,
       any_hits = ifelse(Geneid %in% pir_num_uniq_hits$geneid, 'Yes', 'No'))

boxplot(log1p(total_counts) ~ any_hits , data = pir_total_counts_BlastHits)
```

Huh, for the pirs they're about the same? May be an artefact of generally poor pir coverage.

```{r high_count_unassembled}
pir_total_counts_BlastHits %>% 
  group_by(any_hits) %>% 
  dplyr::slice_max(total_counts, n = 5)
```

```{r low_count_assembled}
pir_total_counts_BlastHits %>% 
  group_by(any_hits) %>% 
  dplyr::slice_min(total_counts, n = 5) %>% 
  map_df(rev) #Turn df upside-down
```

The ancestral _pir_ _PCHAS_0101200_ is *not* detected in the Trinity assembly!
This is a major omission. Should BLAST the Trinity assembly just against this gene.
Use 'PCHAS_1100300' as a control to check that using Trinity as a BLASTdb works.
