---
title: "pirDeNovo_process"
output: 
  html_document:
    theme: cosmo
    code_download: true
    toc: true
    number_sections: true
    toc_float:
      collapsed: false
      smooth_scroll: true
---

# Intro

This script takes the filtered _de novo_ transcripts from _P. vivax_ (filtered for transcripts that match _pir_s (HMM coverage > 75%), and are expressed (Both TPM and Number of Reads greater than 1 in at least one sample).

Different experiments are denoted by an experiment code of the following structure: 
* (Lifecycle stage)(Geographical Source)(Year of publication).
E.g. AsexCamb19 = Aexual stages, Cambodia, 2019. MixdBraz22 = Mixed stages, Brazil, 2022.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, 
                      results = FALSE, 
                      message = FALSE, 
                      warning = FALSE,
                      fig.show = FALSE)
# Default hide output in knitr
options(scipen = 999) 
# Means that numbers are displayed normally not as 1e6 (1000000)
```

```{r loading_packages, include = FALSE, message = FALSE}
# Load the required packages
library('readr')
library('dplyr')
library('stringr')
library('purrr')
library('kableExtra')
library('readxl')
library('rmarkdown')
library('seqinr')
library('ggplot2')
library('ComplexHeatmap')

safe_colorblind_palette <- c("#88CCEE", "#CC6677", "#DDCC77", "#117733", "#332288", "#AA4499", 
                             "#44AA99", "#999933", "#882255", "#661100", "#6699CC", "#888888")

#Set global_save_var as TRUE to save plots.
global_save_var <- FALSE
```

```{r import_funs}
source('../rfunctions/read_HMMertxtgz_fun.R')
source('../rfunctions/uniquify_fun.R')
source('../rfunctions/ordering_fun.R')
```

```{r load_assembly_info}
assembly_stats.df <- read_xlsx('../suppFiles/AdditionalFile1_samples.xlsx') %>% 
  mutate(helpful_id = paste0(experiment_id, '_', helpful_assembly_name),
         country = str_extract(location, pattern = '(?<=_).+(?=_)')) %>%
  select(helpful_id,
         helpful_assembly_name, 
         num_pir, 
         lifecycle_stage, 
         location, 
         experiment_id,
         country,
         BUSCO_Complete) %>% 
  filter(!is.na(num_pir))
```

```{r load_pir_quant}

pir.df <- read.csv('../suppFiles/AdditionalFile3_pirClustersSubfam.csv', 
                   row.names = 1,
                   stringsAsFactors = FALSE)

```

```{r fig3}

subfam_col <- safe_colorblind_palette
names(subfam_col) <- c('E','G','K','J','I','B','C')

( 
  subfam_graph.df <- pir.df %>% 
    arrange(BLAST_Cluster_Num) %>% 
    group_by(BLAST_Cluster_Num) %>% 
    filter(BLAST_Cluster_Num < 26) %>% 
    mutate(infl_net = factor(paste0(BLAST_Cluster_Num, 
                             ' (n = ', n_distinct(label), ')'))) %>% 
    filter(str_detect(Subfam_Sal1, pattern = '^(B|C|D|E|F|G|H|I|J|K)$')) %>%
    group_by(Subfam_Sal1, species, infl_net, .drop = FALSE) %>% 
    dplyr::summarise(number = n_distinct(label))
)

max_y <- max(summarise(group_by(subfam_graph.df, 
                                infl_net), 
                       sum_num = sum(number))$sum_num)
breaks_y <- seq(0, 
                plyr::round_any(max_y, 5, ceiling), 
                5)

( subfam_graph.plot <- ggplot(subfam_graph.df, 
                              aes(x = infl_net, 
                                  y = number, 
                                  fill = Subfam_Sal1)) +
    geom_col(position = 'stack') +
    xlab('Cluster ID (total number of pirs in cluster)') +
    ylab('Number of Sal1 pirs in each cluster') +
    scale_fill_manual(values = subfam_col,
                      name = 'Sal1 sub-family') +
    scale_y_continuous(breaks = breaks_y,
                       # limits = c(0, max(breaks_y))
    ) +
    theme_classic()+
    theme(axis.text.x = element_text(angle = 75, vjust = 0.47)) )

ggsave('../plots/Fig3_vivax_subfamDisI14Network.pdf',
       height = 5,
       width = 10,
       subfam_graph.plot)

```

# Assembly statistics - Figure 2

Plots of the number of _de novo pir_\s vs quality score (BUSCOs - percentage of conserved orthologs expected in the assemblies that were actually found), separated into mosquito and human-infection parasite stages.

Mosquito stages show low-to-no _de novo pir_\s even with decent BUSCO results, with the exception of MosqPeru23 results which included fresher blood meals.

Asexual stages show that better assemblies demonstrate better _de novo pir_ recoveries. There is no plateau of the _pir_ numbers, demonstrating that these assemblies are only capturing a subset of the total _pir_ transcriptome, as expected. Note that one SchzMixd22 (from Papua New Guinea) sample shows exceptional _de novo pir_ numbers (373) despite low BUSCO score.

```{r fig2_assembly_stats}
assembly.df <- unique(select(assembly_stats.df, 
                             lifecycle_stage, BUSCO_Complete, num_pir, experiment_id))

# Keep the same axes for both graphs
y_max <- max(assembly.df$num_pir)

lapply(c('mosquito','asexual'), function(stages){
  busco.plot <- ggplot(data = filter(assembly.df,
                                     str_detect(lifecycle_stage, 
                                                pattern = stages)) , 
                       aes(x = BUSCO_Complete, 
                           y = num_pir, 
                           col = experiment_id,
                           shape = lifecycle_stage)) +
    geom_point(size = 3, stroke  = 2) +
    scale_shape_manual(values = seq(1,10)) +
    xlab('Percentage of complete BUSCOs') +
    ylab('Number of pirs') +
    theme_classic() +
    # Minus one minimum otherwise the zero values are lost.
    ylim(min = -1, max = y_max) +
    xlim(min = 0, max = 100) +
    guides(color = guide_legend(ncol = 1,
                                title = 'Experiment ID', 
                                byrow = TRUE,
                                order = 1),
           shape = guide_legend(ncol = 1, 
                                title = 'Lifecycle stage', 
                                byrow = TRUE)) +
    theme(legend.text = element_text(size = 15),
          legend.key.size = unit(.5,"cm"),
          legend.spacing.y = unit(.25, 'cm'),
          text = element_text(size = 17))
  # ggsave(paste0('../plots/Fig2_vivax_BUSCOvNumPir_', stages,'.pdf'),
  #        busco.plot,
  #        width = unit(10, "in"),
  #        height = unit(6, "in"))
  busco.plot
})

```

# Sub-family distributions across different sources of _pir_ \s.

## Fig 5: Sub-family distributions across the **numbers** of _pir_ \s from different **species**.

When generating the BLAST similarity network I included _pir_\s from other closely related species as well as the Sal1, P01 and W1 _P. vivax_ reference genomes. This heatmap shows the proportion of the total _pir_\s  for each species/vivax genome that belong to each sub-family. 

Note that the colour scale is skewed to show low proportions, as otherwise the genomes dominated by only one or two sub-families make the heatmap only yellow and blue.

_P. knowlesi_ and _coatneyi_ are the most divergent, only showing a few overlapping _pir_ sub-families with the other species. The _P. vivax_ genomes, the _de novo P. vivax_ sequences, and the _P. vivax-like_ and _P. cynomolgi_ genomes contain most of the sub-families defined. 

Interestingly by this metric the hierarchical clustering of the heatmap separates the _de novo_ sequences out with Sal1 and _P. vivax-like_. I'm not sure what to read into this, as it isn't clear why the clustering has done that. Note that the _de novo_ sequences do not include sub-family B, but _pir_\s of this ilk are indeed present in all three reference genomes. This is likely an un/lowly-expressed sub-family.

```{r combine_df}
#Combine the data frames for later figures.
pir_assembly_stats.df <- full_join(pir.df, filter(assembly_stats.df, num_pir > 0) )
```


```{r subfam_dis_spec, fig.show = TRUE}

#Save the species of Plasmodium but save the vvx species as the strain, and Sal1 sequences as 'vvxSal1'
assigned_subfam_num.df <- mutate(pir_assembly_stats.df,
                                 species = ifelse(species == 'vvx' & !str_detect(Subfam_Sal1, 
                                                                                 pattern = '^(B|C|D|E|F|G|H|I|J|K)$'),
                                                  Subfam_Sal1,
                                                  ifelse(species == 'vvx' & str_detect(Subfam_Sal1,
                                                                                       pattern = '^(B|C|D|E|F|G|H|I|J|K)$'),
                                                         'vvxSal1',
                                                         species))) %>% 
  #Filter the de_novo pirs by the BUSCO scores
  filter(is.na(BUSCO_Complete) | BUSCO_Complete >= 50) %>% 
  group_by(species, BLAST_Cluster_Num, Subfam_new) %>% 
  summarise(num_genes = n_distinct(label))

assigned_subfam_prop.df <- group_by(assigned_subfam_num.df, 
                                    species) %>% 
  arrange(BLAST_Cluster_Num) %>% 
  mutate(prop_genes = num_genes/sum(num_genes))

# To order the sub-families from largest to smallest in the heatmap.
assigned_subfam_prop.df$new_subfam_uniq = factor(assigned_subfam_prop.df$Subfam_new, 
                                                 levels = unique(assigned_subfam_prop.df$Subfam_new))

assigned_subfam_prop.mat <- reshape2::dcast(assigned_subfam_prop.df, 
                                            new_subfam_uniq ~ species, 
                                            value.var = 'prop_genes')
rownames(assigned_subfam_prop.mat) <- assigned_subfam_prop.mat$new_subfam_uniq
assigned_subfam_prop.mat$new_subfam_uniq <- NULL

assigned_subfam_prop.mat[is.na(assigned_subfam_prop.mat)] <- 0
# max_col <- max(na.omit(assigned_subfam_prop.mat))
(spec_ht <- Heatmap(t(assigned_subfam_prop.mat[1:26,]), 
                    na_col = 'white',
                    # cluster_rows = FALSE,
                    # cluster_columns = FALSE,
                    row_split = 4,
                    # column_split = 3,
                    row_names_gp = gpar(fontsize = 12),
                    column_names_gp = gpar(fontsize = 14),
                    column_names_rot = 55,
                    col = circlize::colorRamp2(c(0,0.0001,
                                                 max(assigned_subfam_prop.mat)/8,
                                                 max(assigned_subfam_prop.mat)),
                                               c('white',viridis::cividis(3))),
                    heatmap_legend_param  = list(title_gp = gpar(fontsize = 14)),
                    name = 'Proportion of\n pir numbers\n',
) )

# pdf("../../plots/Fig5_vvx_Subfam-Propheatmap_Species.pdf", width = 8, height = 6)
# draw(spec_ht)
# dev.off()
```

## SuppFig3: Sub-family distributions across the **numbers** of _pir_\s from different **assemblies**.

Now I'm focussing on the individual assemblies, defined by their experiment ID (see the beginning of this doc) and an assembly ID after the underscore.

Assemblies were included if they had >= 50% complete BUSCO score (as a quality filter), except for the exceptional PNG _pir_ sample which was also included.

The BUSCO score and the country of origin is included beside the main heatmap. 

I think the heatmap is still mostly splitting based on the number of _pir_ sub-families found in the data, and this is roughly correlated with the group of assemblies with lower BUSCO scores (the bottom 'chunk').

Beyond that, finding patterns is tricky. Certainly looks to me like most samples, even from very different regions, tend to have transcripts for a range of the same sub-families.

```{r}
filter(pir_assembly_stats.df, Subfam_Sal1 == 'de_novo') %>% 
  filter(assembly == 'Meul22_PNG056_vvx' | BUSCO_Complete >= 50) %>%
  group_by(expt, helpful_id, country, assembly_id, BLAST_Cluster_Num, Subfam_new, BUSCO_Complete) %>% 
  summarise(num_genes = n_distinct(label))
```


```{r SuppFig3_subfam_dis_expt, show.fig = TRUE}
assigned_subfam_num.df <- assigned_subfam.df %>% 
  filter(subfam == 'de_novo') %>% 
  filter(assembly == 'Meul22_PNG056_vvx' | Complete >= 50) %>%
  group_by(expt, helpful_id, country, assembly_id, I1.4, new_subfam_uniq, Complete) %>% 
  summarise(num_genes = n_distinct(label))

# ggplot(assigned_subfam_num.df, aes(x = helpful_id, y = num_genes, fill = new_subfam_uniq)) +
#          geom_col(position = 'fill')

assigned_subfam_prop.df <- group_by(assigned_subfam_num.df, 
                                    helpful_id) %>% 
  arrange(I1.4) %>% 
  mutate(prop_genes = num_genes/sum(num_genes))

# To order the sub-families from largest to smallest in the heatmap.
assigned_subfam_prop.df$new_subfam_uniq = factor(assigned_subfam_prop.df$new_subfam_uniq, 
                                                 levels = unique(assigned_subfam_prop.df$new_subfam_uniq))

assigned_subfam_prop.mat <- reshape2::dcast(assigned_subfam_prop.df, 
                                            new_subfam_uniq ~ helpful_id, 
                                            value.var = 'prop_genes')
rownames(assigned_subfam_prop.mat) <- assigned_subfam_prop.mat$new_subfam_uniq
assigned_subfam_prop.mat$new_subfam_uniq <- NULL

assigned_subfam_prop.mat[is.na(assigned_subfam_prop.mat)] <- 0
# max_col <- max(na.omit(assigned_subfam_prop.mat))
subfam_ht <- Heatmap(t(assigned_subfam_prop.mat)[sort(colnames(assigned_subfam_prop.mat)),], 
                     na_col = 'white',
                     name = 'Prop Sub-fam',
                     # cluster_rows = FALSE,
                     # cluster_columns = FALSE,
                     row_names_gp = gpar(fontsize = 6),
                     column_names_gp = gpar(fontsize = 6),
                     row_split = 2,
                     # column_split = 4,
                     col = circlize::colorRamp2(c(0,0.0001,
                                                  max(assigned_subfam_prop.mat)/4,
                                                  max(assigned_subfam_prop.mat)),
                                                c('white',viridis::cividis(3))))

# How does this correspond with the BUSCO scores?

busco.mat <- assigned_subfam_prop.df %>% 
  select(helpful_id, Complete) %>% 
  mutate(Complete = as.numeric(Complete)) %>% 
  t 
names_busco.mat <- busco.mat[1,]
busco.mat <- t(matrix(as.numeric(busco.mat[-1,])))
duplicated_val <- !duplicated(names_busco.mat)
busco.mat <- matrix(busco.mat[duplicated_val])
rownames(busco.mat) <- names_busco.mat[duplicated_val]

busco_ht <- Heatmap(busco.mat[sort(rownames(busco.mat)),],
                    name = 'BUSCO',
                    cluster_rows = FALSE,
                    cluster_columns = FALSE,
                    show_column_names = FALSE,
                    col = circlize::colorRamp2(c(min(busco.mat),
                                                 (max(busco.mat) + min(busco.mat))/2,
                                                 max(busco.mat)),
                                               c(viridis::magma(3))))

country.mat <- t(as.matrix(reshape2::dcast(unique(select(assigned_subfam_prop.df, helpful_id, country), 
                                                  by = cols) , 
                                           1 ~ helpful_id,
                                           value.var = 'country')[,-1]))
country_ht <- Heatmap(country.mat,
                      name = 'Source',
                      row_names_gp = gpar(fontsize = 6),
                      column_names_gp = gpar(fontsize = 6),
                      col = c('black', 'blue', 'yellow', 'red', 'grey', 'hotpink')[1:length(unique(country.mat))])

pdf("../../plots/SuppFig3_vvx_Subfam-Propheatmap_orderCluster.pdf", width = 7, height = 5)
draw(subfam_ht + busco_ht + country_ht)
dev.off()
# biplot(princomp(t(assigned_subfam_prop.mat), scale = TRUE, center = TRUE))

```

## Sub-family distributions across the **transcription (TPM)** of _pir_\\s from different **assemblies**.

Similar heatmap to above except now the proportions are proportion of total expression, as measured in the median TPM of each sample that contributed to the assembly.

Note that max transcription was used to define whether a _pir_ was a real signal or not, so it only had to be transcribed above the thresholds described at the top of this document in at least a single sample. Median TPM for many of these _pir_\s is virtually zero, demonstrating transcription only in under half of samples. The expression of sub-families across different samples (the next section) should enlighten this.

Once again expression is found across most sub-families for most samples, no matter the source. The hiararchical clustering does separate a few Peruvian samples together, although the source of this difference is tricky to elucidate. There is some enhanced sub-family G, I_1 and I_2 expression in these samples, but this doesn't hold for all them, let alone all the Peruvian samples.

```{r tpm}
assigned_subfam_tpm.df <- assigned_subfam.df %>% 
  filter(assembly == 'Meul22_PNG056_vvx' | Complete >= 50
         # ,str_detect(new_subfam_uniq, pattern = 'E|I')
  ) %>% 
  select(expt, helpful_id, assembly_id, I1.4, new_subfam_uniq, Complete, TPM_med)

assigned_subfam_tpm_prop.df <- group_by(assigned_subfam_tpm.df, 
                                        helpful_id, 
                                        new_subfam_uniq,
                                        I1.4)  %>% 
  summarise(TPM_med = sum(TPM_med)) %>% 
  group_by(helpful_id) %>% 
  arrange(I1.4) %>% 
  mutate(prop_tpm = TPM_med/sum(TPM_med)) 

assigned_subfam_tpm_prop.mat <- reshape2::dcast(assigned_subfam_tpm_prop.df, 
                                                new_subfam_uniq ~ helpful_id, 
                                                value.var = 'prop_tpm')
rownames(assigned_subfam_tpm_prop.mat) <- assigned_subfam_tpm_prop.mat$new_subfam_uniq
#Remove the subfam column now that they are rownames.
assigned_subfam_tpm_prop.mat$new_subfam_uniq <- NULL
# Set all NA to zero.
assigned_subfam_tpm_prop.mat[is.na(assigned_subfam_tpm_prop.mat)] <- 0

# max_col <- max(na.omit(assigned_subfam_prop.mat))
tpm_ht <- Heatmap(t(assigned_subfam_tpm_prop.mat)[sort(colnames(assigned_subfam_tpm_prop.mat)),], 
                  na_col = 'white',
                  name = 'Prop TPM',
                  # cluster_rows = FALSE,
                  # cluster_columns = FALSE,
                  column_names_gp = gpar(fontsize = 14),
                  column_names_rot = 55,
                  row_split = 4,
                  heatmap_legend_param  = list(title_gp = gpar(fontsize = 12)),
                  # column_split = 4,
                  col = circlize::colorRamp2(c(0,0.0001,
                                               max(assigned_subfam_tpm_prop.mat)/4,
                                               max(assigned_subfam_tpm_prop.mat)),
                                             c('white',viridis::viridis(3))))

# Note that the busco.mat and country.mat are defined a couple of chunks above.
#  I re-define busco_ht and country_ht here to alter font sizes and remove the row names.

busco_ht <- Heatmap(busco.mat[sort(rownames(busco.mat)),],
                    name = 'BUSCO',
                    cluster_rows = FALSE,
                    cluster_columns = FALSE,
                    show_column_names = FALSE,
                    heatmap_legend_param  = list(title_gp = gpar(fontsize = 12)),
                    col = circlize::colorRamp2(c(min(busco.mat),
                                                 (max(busco.mat) + min(busco.mat))/2,
                                                 max(busco.mat)),
                                               c(viridis::magma(3))))

country_ht <- Heatmap(country.mat,
                      name = 'Source',
                      row_names_gp = gpar(fontsize = 0),
                      column_names_gp = gpar(fontsize = 6),
                      heatmap_legend_param  = list(title_gp = gpar(fontsize = 12)),
                      col = c('black', 'blue', 'yellow', 'red', 'grey', 'hotpink')[1:length(unique(country.mat))])

pdf("../../plots/Fig6_vvx_Subfam-TPMheatmap_orderCluster.pdf", width = 7, height = 6)
draw(tpm_ht + busco_ht + country_ht)
dev.off()
```

### Sub-family distributions across the **transcription (TPM)** of _pir_\\s from different **assemblies** - ordered by location.

```{r subfam_tpm_countryOrder_heatmap}
country_order <- names(country.mat[sort(country.mat, index.return = TRUE)$ix,])
fonts <- 8

tpm_country_ht <- Heatmap(t(assigned_subfam_tpm_prop.mat)[country_order,], 
                          na_col = 'white',
                          name = 'Prop TPM',
                          cluster_rows = FALSE,
                          # cluster_columns = FALSE,
                          # row_names_gp = gpar(fontsize = fonts),
                          column_names_gp = gpar(fontsize = fonts),
                          row_split = substr(str_replace(country.mat[sort(country.mat, 
                                                                          index.return = TRUE)$ix], 
                                                         pattern = 'ElSalv', 
                                                         replacement = 'ElSv'), 1,4),
                          row_title_rot = 0,
                          # column_split = 4,
                          col = circlize::colorRamp2(c(0,0.0001,
                                                       max(assigned_subfam_tpm_prop.mat)/4,
                                                       max(assigned_subfam_tpm_prop.mat)),
                                                     c('white',viridis::viridis(3))))

busco_country_ht <- Heatmap(busco.mat[country_order,],
                            name = 'BUSCO',
                            cluster_rows = FALSE,
                            cluster_columns = FALSE,
                            show_column_names = FALSE,
                            col = circlize::colorRamp2(c(min(busco.mat),
                                                         (max(busco.mat) + min(busco.mat))/2,
                                                         max(busco.mat)),
                                                       c(viridis::magma(3))))

country_country_ht <- Heatmap(country.mat[country_order,],
                              name = 'Source',
                              row_names_gp = gpar(fontsize = fonts),
                              show_column_names = FALSE,
                              col = c('black', 'blue', 'yellow', 'red', 'grey', 'hotpink')[1:length(unique(country.mat))])

pdf("../../plots/vvx_Subfam-TPMheatmap_orderCountry.pdf", width = 7, height = 5)
draw(tpm_country_ht + busco_country_ht + country_country_ht)
dev.off()
```

```{r subfam_tpm_buscoOrder_heatmap}
busco_order <- rev(names(busco.mat[sort(busco.mat, index.return = TRUE)$ix,]))

tpm_busco_ht <- Heatmap(t(assigned_subfam_tpm_prop.mat)[busco_order,], 
                        na_col = 'white',
                        name = 'Prop TPM',
                        cluster_rows = FALSE,
                        # cluster_columns = FALSE,
                        row_names_gp = gpar(fontsize = 6),
                        column_names_gp = gpar(fontsize = 6),
                        # column_split = 4,
                        col = circlize::colorRamp2(c(0,0.0001,
                                                     max(assigned_subfam_tpm_prop.mat)/4,
                                                     max(assigned_subfam_tpm_prop.mat)),
                                                   c('white',viridis::viridis(3))))

busco_busco_ht <- Heatmap(busco.mat[busco_order,],
                          name = 'BUSCO',
                          cluster_rows = FALSE,
                          cluster_columns = FALSE,
                          show_column_names = FALSE,
                          col = circlize::colorRamp2(c(min(busco.mat),
                                                       (max(busco.mat) + min(busco.mat))/2,
                                                       max(busco.mat)),
                                                     c(viridis::magma(3))))

country_busco_ht <- Heatmap(country.mat[busco_order,],
                            name = 'Source',
                            row_names_gp = gpar(fontsize = 6),
                            column_names_gp = gpar(fontsize = 6),
                            col = c('black', 'blue', 'yellow', 'red', 'grey', 'hotpink')[1:length(unique(country.mat))])

pdf("../../plots/vvx_Subfam-TPMheatmap_orderBUSCO.pdf", width = 7, height = 5)
draw(tpm_busco_ht + busco_busco_ht + country_busco_ht)
dev.off()
```


```{r country}
assigned_subfam_num.df <- assigned_subfam.df %>% 
  filter(assembly == 'Meul22_PNG056_vvx' | Complete >= 50) %>% 
  group_by(country) %>% 
  mutate(country = paste0(country, ' (n = ', n_distinct(assembly),')')) %>% 
  group_by(country, I1.4, new_subfam_uniq) %>% 
  summarise(num_genes = n_distinct(label)) 

# ggplot(assigned_subfam_num.df, aes(x = country, y = num_genes, fill = new_subfam_uniq)) +
#          geom_col(position = 'fill')

assigned_subfam_prop.df <- group_by(assigned_subfam_num.df, 
                                    country) %>% 
  arrange(I1.4) %>% 
  mutate(prop_genes = num_genes/sum(num_genes))

# To order the sub-families from largest to smallest in the heatmap.
assigned_subfam_prop.df$new_subfam_uniq = factor(assigned_subfam_prop.df$new_subfam_uniq, 
                                                 levels = unique(assigned_subfam_prop.df$new_subfam_uniq))

assigned_subfam_prop.mat <- reshape2::dcast(assigned_subfam_prop.df, 
                                            new_subfam_uniq ~ country, 
                                            fun.aggregate = median,
                                            value.var = 'prop_genes')
rownames(assigned_subfam_prop.mat) <- assigned_subfam_prop.mat$new_subfam_uniq
assigned_subfam_prop.mat$new_subfam_uniq <- NULL

assigned_subfam_prop.mat[is.na(assigned_subfam_prop.mat)] <- 0
# max_col <- max(na.omit(assigned_subfam_prop.mat))
subfam_ht <- Heatmap(t(assigned_subfam_prop.mat)[sort(colnames(assigned_subfam_prop.mat)),], 
                     na_col = 'white',
                     name = 'Prop Sub-fam',
                     # cluster_rows = FALSE,
                     # cluster_columns = FALSE,
                     row_names_gp = gpar(fontsize = 6),
                     column_names_gp = gpar(fontsize = 6),
                     # row_split = 2,
                     # column_split = 4,
                     col = circlize::colorRamp2(c(0,0.0001,
                                                  max(assigned_subfam_prop.mat)/4,
                                                  max(assigned_subfam_prop.mat)),
                                                c('white',viridis::viridis(3))))

pdf("../../plots/vvx_Subfam-TPMheatmap_Country.pdf", width = 7, height = 5)
draw(subfam_ht)
dev.off()

#From limited data perhaps PNG (Both the de novo and the P01) is depleted for the largest sub-families (Es).

```

TODO:
* There's something wrong with the colour scale of the heatmaps, can be seen when changing filter. - to do with max values?
* What is the prop_TPM calculated from? The sum of medians? Why didn't I use the sum of all samples instead of sum of medians?
** Prop for each sample may make more sense for the linear model estimation too.

*Check that the BUSCO scores included haven't led to duplicated transcript names in case of same transcript matching two BUSCO.
