---
title: "tl21-01 chimeric reads"
output: 
  html_document:
    theme: cosmo
    code_download: true
    toc: true
    number_sections: true
    toc_float:
      collapsed: false
      smooth_scroll: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, 
                      results = TRUE, 
                      message = FALSE, 
                      warning = FALSE)
options(scipen = 999) # Means that numbers are displayed normally not as 1e6 (1000000)
```

```{r loading_packages, include = FALSE, message = FALSE}
# Load the required packages
library('readr')
library('dplyr', 
        lib.loc = "/camp/home/littlet/R/x86_64-pc-linux-gnu-library/4.0")
library('stringr')
library('purrr')
library('seqinr')
```

```{r function_ordering}
ordering <- function(to_order, order_vector, fromLast = TRUE){
  order = unlist(sapply(order_vector, 
                        function(x){unique(str_subset(to_order, 
                                                      pattern = paste(x)))}))
  order_missing = unique(to_order[!to_order %in% order])
  unique(c(order, order_missing), fromLast = fromLast)
}
```

```{r load_data}
pir.info <- read.delim('Pchabaudi_pir_info.txt', sep = '\t')
```

```{r read_blast_results}

blast_output_file.list <- list(
  list(blast_output_file =  '/camp/lab/langhornej/working/HPC/littlet/tl21-01_testingTrinityChab24h/blast_pirDatabase/scspades/scspades_SBPrMT-mR-rRNAdep/scspades_SBPrMT-mR-rRNAdep_blastn.outfmt6',
    fasta_file = '/camp/lab/langhornej/working/HPC/littlet/tl21-01_testingTrinityChab24h/scspades/scspades_SBPrMT-mR-rRNAdep/scaffolds.fasta'),
  list(blast_output_file = c('/camp/lab/langhornej/working/HPC/littlet/tl21-01_testingTrinityChab24h/blast_pirDatabase/Trinity/trinitySBPrMT-mR-rRNAdep_rep220125_blastn/trinitySBPrMT-mR-rRNAdep_rep220125_blastn.outfmt6'),
    fasta_file = "/camp/lab/langhornej/working/HPC/littlet/tl21-01_testingTrinityChab24h/trinity-deNovo/trinity_out_dir_SBPrMTcat-miceRemoved-rRNAdepleted_rep220125/Trinity.fasta"),
  list(blast_output_file = c('/camp/lab/langhornej/working/HPC/littlet/tl21-01_testingTrinityChab24h/blast_pirDatabase/rnaspades/rnaspades_SBPrMT-mR-rRNAdep/rnaspades_SBPrMT-mR-rRNAdep_blastn.outfmt6'),
    fasta_file = "/camp/lab/langhornej/working/HPC/littlet/tl21-01_testingTrinityChab24h/rnaspades/rnaspades_SBPrMT-mR-rRNAdep/transcripts.fasta")
)

lapply(blast_output_file.list, 
       function(blasts_element){
         
         blast_output_file <- blasts_element$blast_output_file
         fasta_file  <- blasts_element$fasta_file
         
         prefix <- str_extract(
           last(unlist(str_split(blast_output_file, 
                                 pattern = '/'))), 
           pattern = '.+(?=_blastn\\.outfmt6$)')
         
         #Set a filter for bp length to assume is a valid match
         filter_nident <- 100
         
         outfmt6_headers <- 
           'qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp slen nident' %>% 
           strsplit(split = ' ') %>% 
           unlist
         
         blast <- read.table(blast_output_file,
                             col.names = outfmt6_headers) %>% 
           mutate(geneid = str_extract(sseqid, 
                                       pattern = '(?<=gene=).+(?=___or)'),
                  scovhsp = round( (nident/slen * 100), digits = 2)
           ) %>% 
           filter(nident > filter_nident)
         
         transcripts_chimeric <- blast %>% 
           group_by(sseqid) %>% 
           filter(scovhsp >= 95, 
                  qcovhsp < 50) %>% 
           arrange(geneid)
         
         transcripts_chimeric_allBlast <- blast %>% 
           filter(qseqid %in% transcripts_chimeric$qseqid)
         
         write.table(transcripts_chimeric_allBlast,
                     file = paste0(prefix, '_chimericReads.txt'),
                     sep = '\t', 
                     row.names = FALSE)
         fst <- read.fasta(fasta_file)
         chimeric_fst <- lapply(transcripts_chimeric$qseqid, 
                                function(x) getElement(fst, x))
         write.fasta(chimeric_fst, 
                     names = transcripts_chimeric$qseqid,
                     file.out = paste0(prefix, '_chimericReads.fasta'))
       })
```
