---
title: "tl23-01_chabaudiBLASTanalysis"
output: 
  html_document:
    theme: cosmo
    code_download: true
    toc: true
    number_sections: true
    toc_float:
      collapsed: false
      smooth_scroll: true
---

# Intro

This script is looks at the BLAST results of a _de novo_ assembled transcriptome. 

# What this report shows:

*The level of counts / TPM of pirs that are detected by the Trinity de novo assembled transcriptome.

**e.g. What is the minimum expressed gene detected / What is the maximally expressed gene not detected.

# To-do:

*Check how they BLAST to the rest of the transcriptome, do pir de novo transcripts also BLAST to non-pir transcripts?

*Select the most informative data and poo-poo the rest.

*Save useful files made by this script as csv.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, 
                      results = TRUE, 
                      message = TRUE, 
                      warning = TRUE)
options(scipen = 999) # Means that numbers are displayed normally not as 1e6 (1000000)
```

```{r loading_packages, include = FALSE, message = FALSE}
# Load the required packages
library('readr')
library('dplyr', 
        # lib.loc = "/nemo/home/littlet/R/x86_64-pc-linux-gnu-library/4.0"
        )
library('stringr')
library('purrr')
# library('IRanges')
library('kableExtra')
library('plotly')
# library('seqinr')
library('rmarkdown')
library('data.table')
# library('xlsx')
library('ComplexHeatmap')
library('circlize')
library('viridis')
```

```{r function_import}
sapply(list.files('../rfunctions', full.names = TRUE), source)
```

```{r function_alternateReverse}
alternateReverse <- function(input.vec){
  #Alternate the indexes of a vector, with one alternation in reverse order.
  # Good for colour systems so the end of a gradient is shifted between very different colours.
  output.vec <- c(rbind(input.vec[seq_along(input.vec)%%2==1], 
                        rev(input.vec[seq_along(input.vec)%%2==0])))
  if(length(output.vec) != length(input.vec)){
    #If input.vec is odd length rbind will add an extra value at the end.
    # So the length will be longer, this removes the last, erroneous element.
    return(output.vec[-last(output.vec)])
  }else{
    return(output.vec)
  }
}
```

```{r load_data}
pir.info <- read.delim('Pchabaudi_pir_info.txt', sep = '\t')

counts <- read.csv('tl19-07_twentyfour_counts_samples.csv',
                   header = TRUE,
                   stringsAsFactors = FALSE) %>% 
  filter(Geneid %in% pir.info$Gene.ID) %>% 
  select(colnames(.)[order(colnames(.))])

```

```{r filter_nident}
#Set a filter for bp length to assume is a valid match
filter_nident <- 100
```

Filtering matches by a minimum of `r filter_nident`bp to be a valid match.

```{r read_blast_results}
# list.files('../blast', pattern = 'outfmt')

outfmt6_headers <- 
  'qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp slen nident' %>% 
  strsplit(split = ' ') %>% 
  unlist

blast_files <- c(
  list.files(path = "../../blast", 
             pattern = "blastp.outfmt6",
             full.names = TRUE),
  list.files(path = "../../blast/chabaudi", 
             pattern = "evigene",
             full.names = TRUE)
)

blast <- lapply(blast_files,
                function(blast_output_file){

                  blast_name <- str_extract(blast_output_file,
                                            pattern = '(?<=\\/)21659_.+(?=(2pir|_blast))') %>%
                    str_replace_all(pattern = '#', replacement = '_') %>%
                    str_remove_all(pattern = 'chabaudi_')

                  sample_name <- str_extract(blast_output_file,
                                             pattern = '(?<=\\/)21659_8.[[:digit:]]{1,2}') %>%
                    str_replace_all(pattern = '#', replacement = '_') %>%
                    str_remove_all(pattern = 'chabaudi_')

                  assembler_name <- str_extract(blast_output_file,
                            pattern = 'spadesSC|spadesRNA|trinity|evigene|multi')

                  cdhit_thresh <- case_when(
                    grepl(blast_output_file, pattern = 'cdhit') ~ "No Dup",
                    grepl(blast_output_file, pattern = 'evigene') ~ "No Dup",
                    .default = 'All Seq'
                  )

                  # cdhit_thresh <- ifelse(grepl(blast_output_file, pattern = 'cdhit'),
                  #                        str_extract(blast_output_file,
                  #                                    pattern = 'cdhit(|est)[[:digit:]]{2}') %>%
                  #                          str_replace(pattern = 'cdhitest', replacement = 'cdhit'),
                  #                        'all')

                  read.table(blast_output_file,
                             col.names = outfmt6_headers) %>%
                    mutate(geneid = str_extract(sseqid,
                                                pattern = '.+(?=\\.)'),
                           scovhsp = round( (nident/slen * 100), digits = 2),
                           scovs = round(((abs(send - sstart) + 1)/slen *100), digits = 2),
                           evalue = format(evalue, scientific = TRUE),
                           blast_name = blast_name,
                           sample_name = sample_name,
                           assembler_name = assembler_name,
                           cdhit_thresh = cdhit_thresh
                    ) %>%
                    # Filter for minimum identity and pir gene name.
                    filter(nident > filter_nident,
                           geneid %in% pir.info$Gene.ID) %>%
                    group_by(qseqid) %>%
                    mutate(num_genes_match =  dplyr::n_distinct(geneid))
                }) %>% dplyr::bind_rows()

write_csv(blast,
          path = 'chabaudi_blastp_results.csv')

blast <- read_csv('./chabaudi_blastp_results.csv')

mutate(blast, sample = paste(sample_name, assembler_name, cdhit_thresh, sep = '_')) %>% 
  .$sample %>% 
  unique %>% 
  sort %>% 
  length() == 216

# names(blast) <- str_extract(blast_files, 
#                             pattern = '(?<=\\/)21659_.+(?=(2pir|_blastp))') %>% 
#   str_replace_all(pattern = '#', replacement = '_') %>% 
#   str_remove_all(pattern = 'chabaudi_')
# 
sample_names <- unique(str_extract(blast_files,
                            pattern = '(?<=\\/)21659_8.[[:digit:]]{1,2}')) %>%
  str_replace_all(pattern = '#', replacement = '_') %>%
  str_remove_all(pattern = 'chabaudi_') %>% 
  unique
# 
# assembler_names <- unique(str_extract(blast_files, 
#                             pattern = 'spadesSC|spadesRNA|trinity|evigene'))

```

TODO: Something is wrong with the below, it's producing weird numbers for the multi. - I think this is fixed?

Get the different subject coverage thresholds at >= 95% identity.

```{r cvrg_stats}

# Get the best match for each reference pir gene.
blast_forcov.df <- blast %>%
  group_by(geneid, blast_name) %>%
  #Choosing this by bitscore, scovhsp or pident may be important
  slice_max(order_by = bitscore, n = 1, with_ties = FALSE) %>%
  left_join(mutate(pir.info, geneid = Geneid))
  
#define custom function to calculate min
custom_min <- function(x) {if (length(x)>0) min(x) else NA}
custom_max <- function(x) {if (length(x)>0) max(x) else NA}

transcriptCvrg.df <- lapply(c(50,75,90,95,100), 
                            function(threshold){
                              transcripts_df <- blast_forcov.df %>% 
                                # filter(scovhsp >= threshold,
                                # pident >= 95) %>%
                                group_by(assembler_name, sample_name, cdhit_thresh) %>% 
                                summarise(num_transcripts = n_distinct(sseqid[scovhsp >= threshold &
                                                                                pident >= 95]),
                                          min_test = custom_min(scovhsp[scovhsp >= threshold &
                                                                   pident >= 95]),
                                          num_deNovo = length(qseqid[scovhsp >= threshold &
                                                                       pident >= 95]),
                                          med_mismatch = median(mismatch[scovhsp >= threshold &
                                                                           pident >= 95]),
                                          med_gap = median(gapopen[scovhsp >= threshold &
                                                                     pident >= 95]),
                                          min_scovhsp = custom_min(qcovhsp[scovhsp >= threshold &
                                                                      pident >= 95]),
                                          max_scovhsp = custom_max(qcovhsp[scovhsp >= threshold &
                                                                      pident >= 95])) %>% 
                                mutate(threshold = threshold) 
                            }) %>% bind_rows()

transcriptCvrg.df %>% filter(sample_name == '21659_8_16', 
                             !is.na(cdhit_thresh), 
                             assembler_name == 'multi')
# 
# blast_forcov.df %>% filter(sample_name == '21659_8_16',
#                            !is.na(cdhit_thresh),
#                            assembler_name == 'multi',
#                            pident >= 95,
#                            scovhsp >= 50,
#                            # sseqid == 'PCHAS_1200200.1-p1'
#                            )
  
```

```{r ancestral_gene_coverage}
blast_forcov.df %>% 
  filter(sseqid == 'PCHAS_0101200.1-p1') %>% 
  group_by(sample_name, assembler_name, cdhit_thresh) %>% 
  slice_max(scovhsp) %>% 
  mutate( max(qcovhsp), max(pident)) %>% 
  ungroup() %>% 
  select(sample_name, assembler_name, cdhit_thresh, sseqid, scovhsp, 'max(pident)', 'max(qcovhsp)' ) %>% 
  arrange(scovhsp)
```



```{r coverage_graph}

assembly.col <- c(evigene = 'red',
                   multi = 'blue',
                   spadesSC = 'seagreen' ,
                   spadesRNA = 'limegreen',
                   trinity = 'deeppink')

(gplot <- ggplot(filter(transcriptCvrg.df, threshold %in% c(50,95)), 
                aes(x = assembler_name, 
                    y = num_transcripts,
                    col = assembler_name)) +
  geom_boxplot() +
  theme_classic() +
  theme(axis.text.x = element_blank()) +
  scale_color_manual(values = assembly.col) +
  xlab('Method of transcriptome assembly/meta-assembly') +
  ylab('Number of known transcripts matched') +
  facet_grid(~ threshold + cdhit_thresh, scales = 'free_x'))

# ggsave(filename = '../../plots/chabaudi_numTranscriptsPerAssembByCvrg.pdf',
#        plot = gplot)

```

```{r mismatches}

transcriptCvrg_MultiEvi.df <- filter(transcriptCvrg.df, 
                                     assembler_name == c('evigene') | assembler_name == c('multi') & cdhit_thresh == 'No Dup',
                                     threshold %in% c(50,95))

( mismatch.plot <- ggplot(data = transcriptCvrg_MultiEvi.df,
                          aes(y = med_mismatch,
                              # size = num_deNovo,
                              x = num_transcripts,
                              col = assembler_name,
                              shape = as.factor(threshold))) +
    geom_jitter(alpha = 0.75, size = 2) +
    scale_color_manual(values = assembly.col) +
    # facet_wrap(~ paste(assembler_name, cdhit_thresh, sep = '_')) +
    theme_classic() )

# ggsave(filename = '../../plots/chabaudi_mismatchesPerAssembByCdhit.pdf',
#        plot = mismatch.plot)

( minscovhsp.plot <- ggplot(data = transcriptCvrg_MultiEvi.df,
       aes(y = min_scovhsp,
           # size = num_deNovo,
           x = num_transcripts,
           # col = paste(assembler_name, cdhit_thresh, sep = '_'),
           shape = as.factor(threshold))) +
  geom_jitter(alpha = 0.75, size = 2) +
  facet_wrap(~ paste0(assembler_name)) +
  theme_classic() )

# ggsave(filename = '../../plots/chabaudi_scovhspPerAssembByCdhit.pdf',
#        plot = minscovhsp.plot)

```

# Are the top expressed transcripts found by the assemblies?

```{r}
sample_table <- read.delim('samples_lanes_rMT_updated.list', 
                           header = FALSE,
                           col.names = c('sample','transmission','id') ) %>% 
  mutate(time = str_extract(sample, pattern = '[:digit:]{1,2}(a|p)m|sch'),
         rep = str_extract(sample, pattern = '(?<=-(|m))[[:digit:]]')) %>% 
  mutate(twentyfour = ifelse( str_detect(time, pattern = 'sch'), 
                              time, 
                              paste0(format(strptime(time, "%I%p"), format="%H"), 'h') ) ) %>% 
  mutate(name = paste0(twentyfour,'_',rep),
         sample_name = str_replace(id, pattern = '#', replacement = '_')) %>% 
  select(sample, id, name, sample_name)
```

```{r}
# Note that not all the files in the sample.list are in the expression data.
rmt <- read.csv('tl19-07_twentyfour_tpm_samples.csv') %>% 
  filter(Geneid %in% pir.info$Gene.ID)
rmt <- rmt[,colnames(rmt) %in% c('Geneid', str_subset(colnames(rmt), pattern = 'rMT') )]
rmt <- rmt[,!grepl(colnames(rmt), pattern = 'Unk')]
colnames(rmt) <- unlist(lapply(str_extract_all(colnames(rmt), pattern = 'Geneid|(?<=rMT_)[[:digit:]]{2}h|_[[:digit:]]$|sch'), paste, collapse = ''))
```

```{r}
rmt <- rename(rmt, 'Gene.ID' = 'Geneid') 
rmt <- melt(rmt, id.vars = 'Gene.ID', variable.name = 'name', value.name = 'tpm') 
```


```{r}
# is_empty(setdiff(rmt$name, sample_table$name))
sample_tpm <- merge(sample_table, rmt, by = 'name')
#Need to merge with the blast table but keep the transcripts missed by blast.

transcripts_tpm <- left_join(sample_tpm, 
                             filter(blast_forcov.df, 
                                    cdhit_thresh == 'No Dup'), 
                             by = c('sample_name', 'Gene.ID') ) 


transcripts_tpm <- data.frame(transcripts_tpm) %>% 
  ungroup() %>% 
  tidyr::complete(Gene.ID, 
                  sample_name, 
                  assembler_name,
                  fill = list('pident' = 0,
                              'qcovhsp' = 0,
                              'scovhsp' = 0)) %>% 
  group_by(Gene.ID, sample_name) %>% 
  tidyr::fill(tpm, .direction = "downup") %>% 
  tidyr::drop_na('assembler_name')

# Should be (num pirs * num samples * num assemblers) rows
nrow(transcripts_tpm) == 211*23*5

#Missing pir that now should be present
filter(transcripts_tpm, Gene.ID == 'PCHAS_1468400')

write.csv(transcripts_tpm, file = 'chabaudiAssemblyPirBlast_TPM.csv')
```

```{r}
transcripts_tpm.plot <- transcripts_tpm %>% 
  filter(assembler_name == 'evigene', 
         # pident >= 95
         )

nrow(transcripts_tpm.plot) == 211*23

lapply(unique(transcripts_tpm.plot$sample_name), 
       function(id_sample){
         ggplot(data = filter(transcripts_tpm.plot, sample_name == id_sample) ) +
           geom_point(aes(x = tpm, y = scovhsp, color = pident)) +
           scale_color_continuous(type = 'viridis') +
           ggtitle(id_sample) +
           theme_classic()
       })
```

```{r altogether}
scovhsp_v_tpm <- ggplot(data = transcripts_tpm.plot ) +
  geom_point(aes(x = tpm, y = scovhsp, color = pident)) +
  scale_color_continuous(type = 'viridis') +
  theme_classic()
ggsave(plot = scovhsp_v_tpm, filename = 'scovhsp_v_tpm.png')
scovhsp_v_tpm
```

# BUSCO scores

Removing '21659_8_19' from the results since it's BUSCO score is particularly low and skews the plot.

```{r busco}
#Get the BUSCO files
busco.files <- c(
  list.files(path = '../../nextflow/chab_results_noNumberAtEnd/busco/',
           pattern = "-plasmodium_odb10-busco.batch_summary.txt",
           full.names = TRUE),
  list.files(path = '../../../tl22-03_nfCoreTranscriptCorral/chab_test_results/busco/',
             pattern = "-plasmodium_odb10-busco.batch_summary.txt",
             full.names = TRUE),
  Sys.glob(file.path('../../nextflow','*ORF*','busco','*-plasmodium_odb10-busco.batch_summary.txt'))
)

(busco.df <- lapply(busco.files, function(busco.file) {
  
  sample_name <- str_extract(busco.file,
                             pattern = '(?<=\\/)21659_8.[[:digit:]]{1,2}') %>%
    str_replace_all(pattern = '#', replacement = '_') %>%
    str_remove_all(pattern = 'chabaudi_')
  
  assembler_name <- str_extract(busco.file,
                                pattern = 'spadesSC|spadesRNA|trinity|multi')
  assembler_name <- ifelse(is.na(assembler_name),
                           ifelse(grepl(busco.file, pattern = 'tl22-03'),
                                  'evigene',
                                  'multi'),
                           assembler_name)
  
  cdhit_thresh <- case_when(
                    grepl(busco.file, pattern = 'cdhit') ~ "No Dup",
                    grepl(busco.file, pattern = 'tl22-03_nfCoreTranscriptCorral') ~ "No Dup",
                    .default = 'All Seq'
                  )
  
  read.delim( busco.file, sep = '\t') %>%  mutate(filename = busco.file,
                                                  sample_name = sample_name,
                                                  assembler_name = assembler_name,
                                                  cdhit_thresh = cdhit_thresh)
}) %>% 
    bind_rows %>% 
    arrange(desc(Complete))
)

(busco.plot <- ggplot(filter(busco.df, sample_name != '21659_8_19'), 
                      aes(x = assembler_name, 
                          y = Complete,
                          col = assembler_name)) +
    geom_boxplot() +
    theme_classic() +
    theme(axis.text.x = element_blank(),
          axis.title.y = element_text(size = 8)) +
    scale_color_manual(values = assembly.col, name = 'Assembly') +
    xlab('Method of transcriptome assembly/meta-assembly') +
    ylab('Number of Complete Plasmodiium BUSCOs found') +
    facet_grid(~ cdhit_thresh, scales = 'free_x'))

(busco_prop.plot <- ggplot(filter(busco.df, sample_name != '21659_8_19'), 
                           aes(x = assembler_name, 
                               y = Duplicated/Complete,
                               col = assembler_name)) +
    geom_boxplot() +
    theme_classic() +
    theme(axis.text.x = element_blank()) +
    scale_color_manual(values = assembly.col, name = 'Assembly') +
    xlab('Method of transcriptome assembly/meta-assembly') +
    ylab('Ratio of duplicate/complete Plasmodiium BUSCOs found') +
    facet_grid(~ cdhit_thresh, scales = 'free_x'))

# ggsave(filename = '../../plots/chabaudi_buscoPerCDHITForAssembly.pdf',
#        plot = busco.plot,
#        height = 3)

# write.xlsx2(
#   file = '../../supp_files/suppfile_busco.xlsx',
#   x = busco.df
# )
# 
# write_csv2(
#   path = '../../supp_files/suppfile_busco.csv',
#   x = busco.df
# )
```

```{r num_transcripts}

fasta.files <- c(
  list.files(path = '../../nextflow/chab_results_noNumberAtEnd/prodigal/',
             pattern = "faa",
             full.names = TRUE),
  list.files(path = '../../../tl22-03_nfCoreTranscriptCorral/chab_test_results/evigene/okayset',
             pattern = "aa",
             full.names = TRUE),
  Sys.glob(file.path('../../nextflow','*ORF*','prodigal','*.faa'))
)
(
  number_transcripts.df <- lapply(fasta.files, 
                                  function(fasta.file) {
                                    number_transcripts <- 
                                      if(grepl(fasta.file, pattern  = '.gz$')){
                                        # Use zless to read the zipped files
                                        system(paste0('zless ',
                                                      fasta.file,
                                                      ' | grep ">" | wc -l'),
                                               intern = TRUE)
                                      } else{ 
                                        system(paste0('cat ',
                                                      fasta.file,
                                                      ' | grep ">" | wc -l'),
                                               intern = TRUE)
                                      }
                                    
                                    sample_name <- str_extract(fasta.file,
                                                               pattern = '(?<=\\/)21659_8.[[:digit:]]{1,2}') %>%
                                      str_replace_all(pattern = '#', replacement = '_') %>%
                                      str_remove_all(pattern = 'chabaudi_')
                                    
                                    assembler_name <- str_extract(fasta.file,
                                                                  pattern = 'spadesSC|spadesRNA|trinity|multi')
                                    assembler_name <- ifelse(is.na(assembler_name),
                                                             ifelse(grepl(fasta.file, pattern = 'tl22-03'),
                                                                    'evigene',
                                                                    'multi'),
                                                             assembler_name)
                                    
                                    cdhit_thresh <- case_when(
                                      grepl(fasta.file, pattern = 'cdhit') ~ "No Dup",
                                      grepl(fasta.file, pattern = 'tl22-03_nfCoreTranscriptCorral') ~ "No Dup",
                                      .default = 'All Seq'
                                    )
                                    
                                    data.frame(number_transcripts = as.numeric(number_transcripts),
                                               filename = fasta.file,
                                               sample_name = sample_name,
                                               assembler_name = assembler_name,
                                               cdhit_thresh = cdhit_thresh)
                                  }
  ) %>% bind_rows %>% arrange(desc(number_transcripts))
)

# write.xlsx2(
#   file = '../../supp_files/suppfile_numTranscriptsChabaudi.xlsx',
#   x = number_transcripts.df
# )
# 
# write_csv2(
#   path = '../../supp_files/suppfile_numTranscriptsChabaudi.csv',
#   x = number_transcripts.df
# )

#Need to set the factor levels for this plot as it is different to the others for some reason.
number_transcripts.df$assembler_name <- factor(number_transcripts.df$assembler_name,
                                               ordering(levels(number_transcripts.df$assembler_name),
                                                        "evigene"))

(numTranscript.plot <- ggplot(number_transcripts.df, 
                              aes(x = assembler_name, 
                                  y = log1p(number_transcripts),
                                  col = assembler_name)) +
    geom_boxplot() +
    theme_classic() +
    theme(axis.text.x = element_blank()) +
    scale_color_manual(values = assembly.col, name = 'Assembly') +
    xlab('Method of transcriptome assembly/meta-assembly') +
    ylab('Log Total number of transcripts + 1') +
    facet_grid(~ cdhit_thresh, scales = 'free_x'))

ggsave(filename = '../../plots/chabaudi_numTranscriptPerCDHITForAssembly.pdf',
       plot = numTranscript.plot,
       height = 3)
```

```{r tpm}
#Need this to find which stages correspond to which samples in the TPM dataset.
times_stages_table <- data.frame(time = c('02h', '05h', '08h', '11h', '14h', '17h','20h','23h','sch'),
                                 stage_diff_count = c('Late-Rings',
                                                      'Rings-Troph-conversion',
                                                      'Early-Trophs',
                                                      'Mid-Trophs',
                                                      'Late-Trophs',
                                                      'Early-Rings',
                                                      'Early-Mid-Rings',
                                                      'Late-Mid-Rings',
                                                      'sch-culture')) %>% 
  mutate(stage_diff_count = make.names(stage_diff_count))
#The sample names and the times they correspond to:
times_stages_table <- read_tsv(file = 'samples_lanes_rMT.list', col_names = FALSE) %>% 
  # Convert to the same time format
  mutate(time = ifelse(str_detect(X1, pattern = 'sch'),
                       str_extract(X1, pattern = '(?<=Pc).+(?=-)'),
                       format(strptime(str_extract(X1, pattern = '(?<=Pc).+(?=-)'), "%I%p"), "%Hh")),
         sample_name = str_replace_all(X3, pattern = '#', replacement = '_')) %>% 
  left_join(times_stages_table, by = c('time')) %>% 
  filter(!is.na(stage_diff_count)) %>% 
  mutate(stage_diff_count = uniquify(stage_diff_count))

tpm_data <- read.csv('tl19-07_twentyfour_tpm_samples.csv',
                   header = TRUE,
                   stringsAsFactors = FALSE) %>% 
  filter(Geneid %in% pir.info$Gene.ID) %>% 
  select(colnames(counts)[order(colnames(counts))]) %>% 
  melt(id.vars = 'Geneid', variable.name = "stage", value.name = "tpm") %>% 
  mutate(geneid = Geneid, 
         time = str_extract(stage, pattern = '(?<=rMT_)[[:digit:]]{2}h(?=_)|sch'),
         stage_diff_count = str_extract(stage, pattern = 'sch.culture_[[:digit:]]$|(?<=h_).+_[[:digit:]]$'))

tpm_time.df <- left_join(tpm_data, times_stages_table, by = c('stage_diff_count', 'time'))
# left_join(blast, tpm_time.df, by = )
blast_tpm <- left_join(blast, tpm_time.df, by = c('geneid', 'sample_name'))

# Lowest TPM, but high quality alignments
blast_tpm %>% 
  filter(pident >= 95, scovhsp >= 95) %>% 
  arrange(tpm) %>% 
  select(qseqid, sseqid, sample_name, stage_diff_count, tpm)

num_pir_over1.df <- tpm_time.df %>% 
  filter(Geneid %in% pir.info$Gene.ID, !is.na(sample_name)) %>% 
  group_by(stage_diff_count, sample_name) %>% 
  summarise(num_pir_over1 = n_distinct(Geneid[tpm >= 1]))

filter(tpm_time.df, is.na(sample_name))

transcriptCvrg_numPoss.df <- left_join(transcriptCvrg.df, num_pir_over1.df, by = 'sample_name') %>% 
  #Remove the unknown sample
  filter(sample_name != '21659_8_24') %>% 
  group_by(assembler_name, sample_name, cdhit_thresh, stage_diff_count, threshold) %>% 
  summarise(percentage_poss = round(100*(num_transcripts/num_pir_over1), digits = 2)) 

filter(transcriptCvrg_numPoss.df, threshold == 95) %>% 
  arrange(desc(percentage_poss))

(gtrans_plot <- ggplot(filter(transcriptCvrg_numPoss.df, threshold %in% c(50,95)), 
                aes(x = assembler_name, 
                    y = percentage_poss,
                    col = assembler_name)) +
  geom_boxplot() +
  theme_classic() +
  theme(axis.text.x = element_blank()) +
  scale_color_manual(values = assembly.col, name = 'Assembly') +
  xlab('Method of transcriptome assembly/meta-assembly') +
  ylab('Percentage of known transcripts that are transcribed') +
  facet_grid(~ threshold + cdhit_thresh, scales = 'free_x'))

# ggsave(filename = '../../plots/chabaudi_numTranscribedPerAssembByCvrg.pdf',
#        plot = gtrans_plot,
#        height = 4)
```

```{r subfam}
#Re-do the number of transcripts found by coverage but this time separated by sub-family.
transcriptCvrg_subfam.df <- lapply(c(50,75,90,95,100), 
                            function(threshold){
                              transcripts_df <- blast_forcov.df %>% 
                                # filter(scovhsp >= threshold,
                                # pident >= 95) %>%
                                group_by(assembler_name, sample_name, cdhit_thresh, subfam) %>% 
                                summarise(num_transcripts = n_distinct(sseqid[scovhsp >= threshold &
                                                                                pident >= 95]),
                                          min_test = custom_min(scovhsp[scovhsp >= threshold &
                                                                   pident >= 95]),
                                          num_deNovo = length(qseqid[scovhsp >= threshold &
                                                                       pident >= 95]),
                                          med_mismatch = median(mismatch[scovhsp >= threshold &
                                                                           pident >= 95]),
                                          med_gap = median(gapopen[scovhsp >= threshold &
                                                                     pident >= 95]),
                                          min_scovhsp = custom_min(qcovhsp[scovhsp >= threshold &
                                                                      pident >= 95]),
                                          max_scovhsp = custom_max(qcovhsp[scovhsp >= threshold &
                                                                      pident >= 95])) %>% 
                                mutate(threshold = threshold) 
                            }) %>% bind_rows()

# Join the TPM data with the pir gene information, and summarise how many of each subfam are expressed over 1 TPM.
num_pir_over1_subfam.df <- left_join(tpm_time.df, pir.info, by = 'Geneid') %>% 
  filter(Geneid %in% pir.info$Gene.ID, !is.na(sample_name)) %>% 
  group_by(stage_diff_count, sample_name, subfam, time) %>% 
  summarise(num_pir_over1 = n_distinct(Geneid[tpm >= 1]))

# Combine the above to calculate the percentage of the expressed pirs that are found.
transcriptCvrg_numPoss_subfam.df <- left_join(transcriptCvrg_subfam.df, 
                                              num_pir_over1_subfam.df, 
                                              by = c('subfam','sample_name')) %>% 
  #Remove the unknown sample
  filter(sample_name != '21659_8_24') %>% 
  group_by(assembler_name, sample_name, cdhit_thresh, stage_diff_count, time, threshold, subfam) %>% 
  summarise(percentage_poss = ifelse(num_pir_over1 > 0,
                                     round(100*(num_transcripts/num_pir_over1), digits = 2),
                                     100 + num_transcripts))

lapply(unique(transcriptCvrg_numPoss_subfam.df$assembler_name), function(assembler.nam){
  #Turn into a matrix shape, ordered by assembly, cd_hit threshold and the stage of the asexual cycle.
  # Use forcats::as.factor to ensure the ordering during dcast matches the ordering of the melted data frame.
  transcriptCvrg_numPoss_subfam.mat <- reshape2::dcast(arrange(filter(transcriptCvrg_numPoss_subfam.df, 
                                                                      threshold == 95, 
                                                                      assembler_name == assembler.nam,
                                                                      # subfam != 'ancestral'
                                                                      ), 
                                                               assembler_name, cdhit_thresh, time), 
                                                       forcats::as_factor(paste(
                                                         # assembler_name, 
                                                         sample_name, cdhit_thresh, stage_diff_count, 
                                                         sep = '_')) ~ subfam,
                                                       value.var = 'percentage_poss')
  rownames(transcriptCvrg_numPoss_subfam.mat) <- transcriptCvrg_numPoss_subfam.mat[,1]
  transcriptCvrg_numPoss_subfam.mat <- transcriptCvrg_numPoss_subfam.mat[,-1]
  
  ComplexHeatmap::Heatmap(transcriptCvrg_numPoss_subfam.mat, 
                          cluster_rows = FALSE,
                          cluster_columns = FALSE,
                          na_col = 'white',
                          row_names_gp = gpar(fontsize = 5),
                          col =  colorRamp2(c(0, 50, 100)/4, viridis(3)))
})

tmp <- left_join(transcriptCvrg.df, 
                 summarise(group_by(num_pir_over1_subfam.df,
                                    stage_diff_count, sample_name, time),
                           num_pir_over1 = sum(num_pir_over1)), 
                 by = c('sample_name'))
lapply(unique(tmp$threshold), function(threshold.val){
  tmp <- filter(tmp, threshold == threshold.val)
  plot(tmp$num_pir_over1, tmp$num_transcripts, main = threshold.val)
})

```


```{r}
# if(any_blast){
  number_knownCovOver95 <- length(unique(transcripts_covered_by_denovo$geneid))
  number_knownCovOver75 <- length(unique(transcripts_covered_by_denovo_75$geneid))
  
  number_knownCovOver95_idOver95 <-length(unique(filter(transcripts_covered_by_denovo, 
                                                        pident >= 95)$geneid))
  number_knownCovOver75_idOver95 <- length(unique(filter(transcripts_covered_by_denovo_75, 
                                                         pident >= 95)$geneid))
  
  
  number_knownCovOver95_deNovoCovUnder50 <- length(unique(filter(
    ungroup(transcripts_covered_by_denovo), 
    qcovhsp < 50
  )$geneid))
  
  number_knownCovOver95_idOver95_deNovoCovUnder50 <- length(unique(filter(
    ungroup(transcripts_covered_by_denovo), 
    qcovhsp < 50,
    pident >= 95
  )$geneid))
  
  ancestral_over95_id <- ifelse(
    any(grepl(transcripts_covered_by_denovo$sseqid,
              pattern = 'PCHAS_0101200')),
    paste(transcripts_covered_by_denovo$pident[
      grepl(transcripts_covered_by_denovo$sseqid,
            pattern = 'PCHAS_0101200')
    ], sep = '; '),
    0)
  ancestral_over75_id <- ifelse(
    any(grepl(transcripts_covered_by_denovo_75$sseqid,
              pattern = 'PCHAS_0101200')),
    paste(transcripts_covered_by_denovo_75$pident[
      grepl(transcripts_covered_by_denovo_75$sseqid,
            pattern = 'PCHAS_0101200')
    ], sep = '; '),
    0)
  
  mean_mismatches = mean(transcripts_covered_by_denovo$mismatch)
  mean_gaps = mean(transcripts_covered_by_denovo$gapopen)
  med_mismatches = median(transcripts_covered_by_denovo$mismatch)
  med_gaps = median(transcripts_covered_by_denovo$gapopen)
# }else{
#   number_knownCovOver95 <- number_knownCovOver95_deNovoCovUnder50 <- 
#     ancestral_over95 <- mean_mismatches <- 
#     mean_gaps <- med_mismatches <- med_gaps <- 0
# }
```

```{r data_save}
df_save <- data.frame(
  "assembly" = params$assembly_name,
  "num_transcripts" = number_transcripts,
  
  "knwn_trnscrpt_cvg_over75" = number_knownCovOver75,
  "knwn_trnscrpt_cvg_over75_idOver95" = number_knownCovOver75_idOver95,
  "ancestral_id_over75" = ancestral_over75_id,
  "knwn_trnscrpt_cvg_over95" = number_knownCovOver95,
  "knwn_trnscrpt_cvg_over95_idOver95" = number_knownCovOver95_idOver95,
  "knwn_trnscrpt_cvg_over95_deNovoCvgUnder50" = number_knownCovOver95_deNovoCovUnder50,
  "ancestral_id_over95" = ancestral_over95_id,
  
  "mean_mismatches" = mean_mismatches,
  "mean_gaps" = mean_gaps,
  "med_mismatches" = med_mismatches,
  "med_gaps" = med_gaps,
  
  "assembly_file" = params$assembly_file,
  "blast_file" = params$blast_output_file
) 

if(!file.exists('/camp/lab/langhornej/working/HPC/littlet/tl21-01_testingTrinityChab24h/tl21-01_assembly_info.txt') | params$new_table){
  #If table file doesn't exist then create it.
  write.table(df_save,
              file = '/camp/lab/langhornej/working/HPC/littlet/tl21-01_testingTrinityChab24h/tl21-01_assembly_info.txt',
              sep = '\t',
              col.names = TRUE,
              append = FALSE,
              row.names = FALSE)
} else if (params$assembly_name %in% read.delim('/camp/lab/langhornej/working/HPC/littlet/tl21-01_testingTrinityChab24h/tl21-01_assembly_info.txt')$assembly){
  #If assembly_name is already in table then overwrite it.
  assembly_info <- read.delim('/camp/lab/langhornej/working/HPC/littlet/tl21-01_testingTrinityChab24h/tl21-01_assembly_info.txt')
  assembly_info[assembly_info$assembly %in% params$assembly_name,] <- df_save[1,]
  write.table(assembly_info,
              file = '/camp/lab/langhornej/working/HPC/littlet/tl21-01_testingTrinityChab24h/tl21-01_assembly_info.txt',
              sep = '\t',
              row.names = FALSE)
} else {
  #Append to the existing table is both above are false
  write.table(df_save,
              file = '/camp/lab/langhornej/working/HPC/littlet/tl21-01_testingTrinityChab24h/tl21-01_assembly_info.txt',
              sep = '\t',
              col.names = FALSE,
              append = TRUE,
              row.names = FALSE)
}
df_save %>% 
  kbl() %>% kable_styling() %>% scroll_box(width = "100%", height = "400px")
```